### ISSUE TITLE
Angular is changing order of css classes in class attribute

### ISSUE THREAD

**Issue by dragonflypl** [angular/angular#33111]
# 🐞 bug report

### Affected Package

One of those, responsible for rendering:

```
    "@angular/animations": "~9.0.0-next.10",
    "@angular/common": "~9.0.0-next.10",
    "@angular/compiler": "~9.0.0-next.10",
    "@angular/core": "~9.0.0-next.10",
    "@angular/forms": "~9.0.0-next.10",
    "@angular/platform-browser": "~9.0.0-next.10",
    "@angular/platform-browser-dynamic": "~9.0.0-next.10",
    "@angular/router": "~9.0.0-next.10",
```

### Is this a regression?

Yes. The application generated with the latest stable version 8 does not have this bug.

### Description

I've started using Angular with Semantic UI.

Some features of Semantic UI rely on order of css classes in class attribute. However Angular is changing order of css classes in class attribute while rendering. E.g. from

```
      <div class="ui left icon input">
        <input type="text" placeholder="Search users..." />
        <i class="users icon"></i>
      </div>
```

I get:

```
<div class="icon input left ui"><input type="text" placeholder="Search users..."><i class="icon users"></i></div>
```

`ui left icon input` was changed to `icon input left ui`. As a result icon is incorrectly positioned.

## 🔬 Minimal Reproduction

Just generate app with `ng new` with version 9 dependencies and put this into the template of the app component

```
      <div class="ui left icon input">
        <input type="text" placeholder="Search users..." />
        <i class="users icon"></i>
      </div>
```

And then compare order of classes in rendered HTML in the browser

## 🌍  Your Environment

**Angular Version:**
<pre><code>
Angular CLI: 9.0.0-next.9
Node: 10.16.2
OS: win32 x64
Angular: 9.0.0-next.10
... animations, common, compiler, compiler-cli, core, forms
... language-service, platform-browser, platform-browser-dynamic
... router

Package                           Version
-----------------------------------------------------------
@angular-devkit/architect         0.900.0-next.9
@angular-devkit/build-angular     0.900.0-next.9
@angular-devkit/build-optimizer   0.900.0-next.9
@angular-devkit/build-webpack     0.900.0-next.9
@angular-devkit/core              9.0.0-next.9
@angular-devkit/schematics        9.0.0-next.9
@angular/cli                      9.0.0-next.9
@ngtools/webpack                  9.0.0-next.9
@schematics/angular               9.0.0-next.9
@schematics/update                0.900.0-next.9
rxjs                              6.5.3
typescript                        3.5.3
webpack                           4.41.0
</code></pre>

**Comment by dannyskoog** [comment]
It seems that static classes are sorted nowadays, yes. I tried to dig into where it happens and I believe it's here https://github.com/angular/angular/blob/master/packages/core/src/render3/util/styling_utils.ts#L361

**Comment by Airblader** [comment]
I think it's pretty difficult to have a "useful" definition of how classes are ordered giving that they can also be influenced by ngClass etc. Generally speaking order of CSS classes on an element does not matter according to spec, except for attribute matching on the class attribute (is Semantic UI doing that?). Retaining the order in the absence of other influential factors might be possible but beyond that I think it will be tough.

**Comment by lacolaco** [comment]
With v9.0.0 RC.1, enabling Ivy occurs this regression.  Opting-out Ivy with `enableIvy: false` can be a workaround.

Some UI framework libraries depend on the order of classes. And the breaking of CSS is tough to get found. I think this has to be fixed for v9 stable.

**Comment by IgorMinar** [comment]
I can repro this quite easily:

```
<div class="c b a"></div>
    => wrong: <div _ngcontent-pbj-c20="" class="a b c"></div>
<div [class]="'c b a'"></div>
    => wrong: <div _ngcontent-pbj-c20="" class="a b c"></div>
<div [ngClass]="'c b a'"></div>
    => wrong: <div _ngcontent-pbj-c20="" ng-reflect-ng-class="c b a" class="a b c"></div>
<div [ngClass]="['c', 'b', 'a']"></div>
    => wrong: <div _ngcontent-pbj-c20="" ng-reflect-ng-class="c,b,a" class="a b c"></div>
<div [attr.class]="'c b a'"></div>
    => ok: <div _ngcontent-pbj-c20="" class="c b a"></div>
```

**Comment by IgorMinar** [comment]
It should be noted that the change of the ordering of css classes doesn't change how css rules are applied to the DOM, it only changes how the order is reflected in the `class` attribute.

So for example given the previous 5 examples and stylesheet:

```
.a {
  color: red;
}

.b {
  color: blue;
}

.c {
  color: yellow;
}
```

"c" class rules will be applied to all 5 `div`s.

**Comment by mhevery** [comment]
For simple cases where we just have static classes this can be fixed. However the moment there are multiple class bindings the algorithm will have to sort these in order to be performant. @matsko  is already working on a refactoring which will fix this one as well.

**Comment by IgorMinar** [comment]
we don't need to keep the order if there are any class bindings on an element, I think the only scenario where the class order should be preserved is the static case: `<div class="c b a"></div>`.

**Comment by adampoczatek** [comment]
Apologies for reopening This is a big issue for anyone using Semantic UI as it's relying heavily on the class order. In addition to that, Semantic UI often requires specifying the same class name multiple times, for instance:

```
<div class="ui grid">
    <div class="
        sixteen wide mobile
        eight wide tablet
        ten wide computer
        ten wide large screen
        eleven wide widescreen column">
        ...
    </div>
</div>
```

_Whether this is a good or bad approach it's a separate matter._

This wasn't a massive problem until now - with Angular <=8 all HTML elements that don't have any directives would remain the same after compilation (all classes that you specified in the template would be rendered in the same order). With Angular 9, however, **all** html elements are stripped out of the duplicated classes and the order gets changed, no matter if they are static or not. 

I think it's fair to say that Semantic UI is unusable with Angular 9 unless the issues above are resolved, in which case it'd be worth communicating that to the users.

---

From what I can see, there are 9 Semantic UI modules that rely on class order:

* https://github.com/Semantic-Org/Semantic-UI/blob/master/src/definitions/modules/popup.less#L637
* https://github.com/Semantic-Org/Semantic-UI/blob/master/src/definitions/views/item.less#L158
* https://github.com/Semantic-Org/Semantic-UI/blob/master/src/definitions/modules/modal.less#L126
* https://github.com/Semantic-Org/Semantic-UI/blob/master/src/definitions/modules/dropdown.less#L144
* https://github.com/Semantic-Org/Semantic-UI/blob/master/src/definitions/elements/segment.less#L69
* https://github.com/Semantic-Org/Semantic-UI/blob/master/src/definitions/elements/container.less#L128
* https://github.com/Semantic-Org/Semantic-UI/blob/master/src/definitions/collections/grid.less#L46
* https://github.com/Semantic-Org/Semantic-UI/blob/master/src/definitions/views/card.less#L195

**Comment by Splaktar** [comment]
@adampoczatek wow 🤯 that's wild and a little painful 😬, but thank you for posting more details on this. To be more clear, it looks like Semantic UI is unusable with Angular version 9 RCs + Ivy (default) at this time, but it should be usable when opting out of Ivy.

I tried to look for some updates on how this is progressing and I found an internal Jira ticket (FW-1715) but it doesn't have any additional information other than this is being actively worked on as recently as last week.
